### PR TITLE
added entities User, Profile, Message and superclass AuditedEntity 

### PR DESCRIPTION
**Pull request by artemlisikh**

### CODE REVIEW

**Comment by olisikh** [comment]
Я завтра поревьюваю и предложу улучшения, молоток, хорошая работа:)

File: src/main/java/com/geronimo/model/AuditedEntity.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@Setter
+@ToString
+@MappedSuperclass
+abstract class AuditedEntity {
+
+    @NotNull
+    protected Date date_created;
+
+    @NotNull
+    protected Date last_updated;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/AuditedEntity.java]: According to Java code convention, all properties that are not constants must be written in came-case.

e.g. lastUpdated, countOfReblogs, countOfLikes, etc.

File: src/main/java/com/geronimo/model/Message.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @NotNull
+    private Long user_id;
+
+    @NotNull
+    private Long count_of_reblogs;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Message.java]: I think we need to have a dedicated table for reblogs, as we need to track the user who did the reblog somehow. Same for likes, we need to know that this particular use has already liked a message or/and reblogged.

this is a Many-to-Many relationship

File: src/main/java/com/geronimo/model/Message.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    @NotNull
+    private Long count_of_likes;
+
+    @ManyToMany
+    private List<Message> answers = new LinkedList<>();
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Message.java]: This should be OneToMany, as "this one particular message could have N answers".

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class GeronimoDatabaseTest {
+
+    private UserRepository userRepository;
+
+    @Test
+    public void testDatabase() {
+
```
Comment (olisikh) [inline src/test/java/com/geronimo/GeronimoDatabaseTest.java]: the test to check that database is working could be just executing a query "SELECT 1" on that database using an entityManager's native SQL query support. I believe there's something for native query execution.

File: src/main/java/com/geronimo/model/AuditedEntity.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.MappedSuperclass;
+import javax.persistence.SequenceGenerator;
+import javax.validation.constraints.NotNull;
+import java.util.Date;
+
+@Setter
+@ToString
+@MappedSuperclass
+abstract class AuditedEntity {
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/AuditedEntity.java]: Spring Data has a special support for auduting: https://docs.spring.io/spring-data/data-jpa/docs/2.0.x/reference/html/#auditing

But it's better to have it configured like this: https://docs.spring.io/spring-data/data-jpa/docs/2.0.x/reference/html/#jpa.auditing to not have a vendor dependency on entity layer

**Review by olisikh — CHANGES_REQUESTED** [review]

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java (hunk 4 above)
Comment (artemlisikh) [reply src/test/java/com/geronimo/GeronimoDatabaseTest.java]: I used autowired userRepository to check if most things are working as supposed to, but later deleted all code from "testDatabase" method. Though, thanks for that piece of advice

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class GeronimoDatabaseTest {
+
+    private UserRepository userRepository;
+    private EntityManagerFactory factory;
+
+    @Test
+    public void testDatabase() {
```
Comment (olisikh) [inline src/test/java/com/geronimo/GeronimoDatabaseTest.java]: Great test! I suggest you to choose more informative names for tests. e.g. "testDatabaseConnectionEstablished"

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+
+    @Autowired
+    public void setUserRepository(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @Autowired
+    public void setFactory(EntityManagerFactory factory) {
```
Comment (olisikh) [inline src/test/java/com/geronimo/GeronimoDatabaseTest.java]: These setter methods could be generated by IntellijIdea, ALT+INSERT and choose setters, as I can see the name of the method differs, it should be "setEntitiyManagerFactory". Usually setter method names contain "set" and the name of the property you want to set.

File: src/main/java/com/geronimo/model/Profile.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    @Lob
+    private byte[] picture;
+
+    private Date dateOfBirth;
+
+    @Column(name = "date_created")
+    @CreatedDate
+    protected Date dateCreated;
+
+    @Column(name = "last_updated")
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Profile.java]: I believe there's a hibernate naming strategy for columns, e.g. transformation of camel case into snake case. It may look a bit harder than just writing a @Column annotation with a name everywhere, but if you forget to put this annotation you'll get a column with inapropriate name. Here's a link for an example of how to setup NamingStrategy: https://www.petrikainulainen.net/programming/tips-and-tricks/implementing-a-custom-namingstrategy-with-hibernate/

File: src/main/java/com/geronimo/model/Message.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+import java.util.*;
+
+@Data
+@Entity
+@Table(name = "messages")
+@EntityListeners(AuditingEntityListener.class)
+public class Message {
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Message.java]: Why did you remove extension from AuditedEntity? Doesn't it works just to put this listener there and have these dateCreated, lastUpdated fields annotated there?

**Review by olisikh — CHANGES_REQUESTED** [review]

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java (hunk 7 above)
Comment (artemlisikh) [reply src/test/java/com/geronimo/GeronimoDatabaseTest.java]: I used alt+insert and as far as I can see, setter method contains the name of the variable, not class name

File: src/main/java/com/geronimo/model/Profile.java (hunk 8 above)
Comment (artemlisikh) [reply src/main/java/com/geronimo/model/Profile.java]: default strategy is camel case into snake case, right?

File: src/main/java/com/geronimo/model/Message.java (hunk 9 above)
Comment (artemlisikh) [reply src/main/java/com/geronimo/model/Message.java]: my bad

File: src/main/java/com/geronimo/model/AuditedEntity.java [hunk 10]
```diff
@@ -2,24 +2,26 @@
...
+    protected Date dateCreated;
 
-    @NotNull
-    protected Date last_updated;
+    @LastModifiedDate
+    protected Date lastUpdated;
 
-    protected Long version = 1L;
+    @javax.persistence.Version
+    protected Long version;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/AuditedEntity.java]: Version column must be in the AbstractEntity, because this column is used for optimistic locking, it's not connected to auditing

File: src/test/java/com/geronimo/GeronimoDatabaseTest.java (hunk 7 above)
Comment (olisikh) [reply src/test/java/com/geronimo/GeronimoDatabaseTest.java]: right, sorry, I thought you named your property the same way as the class name, then it's ok.

File: src/main/java/com/geronimo/model/User.java [hunk 11]
```diff
@@ -1,15 +1,38 @@
...
+
+    @NotNull
     private String username;
+
+    @NotNull
     private String password;
+
+    @OneToOne(cascade=CascadeType.ALL)
+    @NotNull
+    private Profile profile = new Profile();
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/User.java]: Make this class embedded, we don't need to create a new table for profile. Agreed? 

If you make this embedded, fields from Profile class will be extracted to the User table.

File: src/main/java/com/geronimo/model/AbstractEntity.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+import lombok.Setter;
+
+import javax.persistence.MappedSuperclass;
+
+@Setter
+@MappedSuperclass
+public class AbstractEntity {
+
+    @javax.persistence.Version
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/AbstractEntity.java]: Minor things, but makes code look better, you can add import javax.persistence.Version and use Version annotation without explicit package name.

File: src/main/java/com/geronimo/model/AuditedEntity.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+import javax.persistence.EntityListeners;
+import javax.persistence.MappedSuperclass;
+import java.util.Date;
+
+@Setter
+@ToString
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+abstract class AuditedEntity extends AbstractEntity {
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/AuditedEntity.java]: By the way, what about getters? Setter & Getter annotations maybe? 

Also entities need to be comparable by content so we need equals & hashCode, right?
There's an annotation of lombok for that, you'll need to annotate super classes as well, and the classes that extend them should also have this annotation with (callSuper = true) to take super-class methods into consideration.

Comment (olisikh) [reply src/main/java/com/geronimo/model/AuditedEntity.java]: Data annotation does everything as far as I know: setters, getters, equals&hashcode, toString.

File: src/main/java/com/geronimo/model/Message.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+@Entity
+@Table(name = "messages")
+public class Message extends AuditedEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @NotNull
+    private Long userId;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Message.java]: Why userId? did you mean `private User author;` ?

File: src/main/java/com/geronimo/model/Message.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Set;
+
+@Data
+@Entity
+@Table(name = "messages")
+public class Message extends AuditedEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/Message.java]: You have an id field already in AbstractEntity, right?

File: src/main/java/com/geronimo/model/User.java [hunk 16]
```diff
@@ -1,15 +1,38 @@
...
 
-@Entity(name = "users")
+@Entity
 @Data
+@ToString(exclude = {"following", "messages", "followers"})
+@Table(name = "users")
 public class User {
     @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
     private Long id;
```
Comment (olisikh) [inline src/main/java/com/geronimo/model/User.java]: Same here, we already have id field in super class

**Review by olisikh — CHANGES_REQUESTED** [review]
Another review, maybe didn't find all the things, as you'll see problems when you'll try to create objects with relationships, but probably it's ok for now...

File: src/main/java/com/geronimo/model/User.java (hunk 11 above)
Comment (artemlisikh) [reply src/main/java/com/geronimo/model/User.java]: When I started coding it, it was embedded, then I thought it was a bit too messy to keep it all in one table.

File: src/main/java/com/geronimo/model/Message.java (hunk 15 above)
Comment (artemlisikh) [reply src/main/java/com/geronimo/model/Message.java]: Well, I don't

**Review by olisikh — APPROVED** [review]

**Comment by olisikh** [comment]
Approved this PR, you can merge it back to develop now.
